pkg/catalogv2/oci/oci.go

```diff
@@ -0,0 +1,234 @@
+func Chart(credentialSecret *corev1.Secret, chart *repo.ChartVersion, URL string, clusterRepoSpec v1.RepoSpec) (io.ReadCloser, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	if len(chart.URLs) == 0 {
+		return nil, fmt.Errorf("failed to find chartName %s version %s: %w", chart.Name, chart.Version, validation.NotFound)
+	}
+
+	// Create a new OCIClient
+	ociClient, err := NewClient(URL, clusterRepoSpec, credentialSecret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create an oci client for url %s: %w", URL, err)
+	}
+
+	// Create an oras repository
+	orasRepository, err := ociClient.GetOrasRepository()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create an oci repository for url %s: %w", URL, err)
+	}
+
+	// Download the oci artifact manifest
+	memoryStore := memory.New()
+	manifest, err := oras.Copy(ctx, orasRepository, ociClient.tag, memoryStore, "", oras.DefaultCopyOptions)
+	if err != nil {
+		return nil, fmt.Errorf("unable to oras copy the remote oci artifact %s:%w", URL, err)
+	}
+	// Fetch the maniest blob of the oci artifact
+	manifestblob, err := content.FetchAll(ctx, memoryStore, manifest)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch the manifest blob of %s:%w", URL, err)
+	}
+	var manifestJSON ocispecv1.Manifest
+	err = json.Unmarshal(manifestblob, &manifestJSON)
+	if err != nil {
+		return nil, fmt.Errorf("unable to unmarshal manifest blob of %s:%w", URL, err)
+	}
+
+	// Check if the oci artifact is of type helm config ?
+	if manifest.ArtifactType == registry.ConfigMediaType || manifestJSON.Config.MediaType == registry.ConfigMediaType {
+		// find the layer of chart type and fetch it
+		for _, layer := range manifestJSON.Layers {
+			if layer.MediaType == registry.ChartLayerMediaType {
+
+				chartTar, err := content.FetchAll(ctx, memoryStore, layer)
+				if err != nil {
+					return nil, err
+				}
+
+				return io.NopCloser(bytes.NewBuffer(chartTar)), nil
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("unable to find the required chart tar file for %s", URL)
+}
+
+// GenerateIndex creates an helm repo index from the oci url provided
+// by fetching the repositories and then tags according to the url and
+// then adding the chart entry to the helm repo index using the oras library.
+func GenerateIndex(URL string, credentialSecret *corev1.Secret, clusterRepoSpec v1.RepoSpec) (*repo.IndexFile, error) {
+	logrus.Debugf("Generating index for oci clusterrepo URL %s", URL)
+
+	// Create a new oci client
+	ociClient, err := NewClient(URL, clusterRepoSpec, credentialSecret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create an oci client for url %s: %w", URL, err)
+	}
+
+	// Create an empty helm repo index file
+	indexFile := repo.NewIndexFile()
+
+	var orasRepository *remote.Repository
+	chartName := ""
+
+	// Loop over all the tags and add to helm repo index
+	tagsFunc := func(tags []string) error {
+		for _, tag := range tags {
```

Comment:
SemVer has the `+` character which isn't allowed in registry tags. So, Helm switches to/from `_`. You can find the switch back code [here](https://github.com/helm/helm/blob/e81f6140ddb22bc99a08f7409522a8dbe5338ee3/pkg/registry/util.go#L123-L142). Shouldn't tag handling account for this?

Reply:
In this case, we only pull the OCI artifact and the helm codebase [here](https://github.com/helm/helm/blob/e81f6140ddb22bc99a08f7409522a8dbe5338ee3/pkg/registry/util.go#L123-L142) is only relevant when pushing the helm chart into the OCI artifact since chart versions with `+` can't be pushed into OCI registry.

This makes me wonder why we do replace of `+` with `_` while pulling a chart in helm [codebase](https://github.com/helm/helm/blob/main/pkg/registry/client.go#L322)

Reply:
Done.